queries/orderbook/batch_rewards.sql

```diff
@@ -28,13 +28,136 @@ WITH observed_settlements AS (SELECT
                                WHERE block_deadline >= {{start_block}}
                                  AND block_deadline <= {{end_block}}
                                GROUP BY ss.auction_id),
+-- protocol fees:
+order_surplus AS (
+    SELECT
+        ss.winner as solver,
+        at.auction_id,
+        s.tx_hash,
+        t.order_uid,
+        o.sell_token, -- the total amount the user sends
+        o.buy_token,
+        t.sell_amount,
+        t.buy_amount,
+        oe.surplus_fee as observed_fee,
+        o.kind,
+        CASE
+            WHEN o.kind = 'sell'
+                THEN t.buy_amount - o.buy_amount / (o.sell_amount + o.fee_amount) * t.sell_amount
+            WHEN o.kind = 'buy'
+                THEN (o.sell_amount + o.fee_amount) / o.buy_amount * t.buy_amount - t.sell_amount
+        END AS surplus,
+        CASE
+            WHEN o.kind = 'sell'
+                THEN o.buy_token
+            WHEN o.kind = 'buy'
+                THEN o.sell_token
+        END AS surplus_token
+    FROM settlements s -- links block_number and log_index to tx_from and tx_nonce
+    JOIN auction_transaction at -- links auction_id to tx_from and tx_nonce
+        ON s.tx_from = at.tx_from AND s.tx_nonce = at.tx_nonce
+    JOIN settlement_scores ss -- contains block_deadline
+        ON at.auction_id = ss.auction_id
+    JOIN trades t -- contains traded amounts
+        ON s.block_number = t.block_number -- log_index cannot be checked, does not work correctly with multiple auctions on the same block
+    JOIN orders o -- contains tokens and limit amounts
+        ON t.order_uid = o.uid
+    JOIN order_execution oe -- contains surplus fee
+        ON t.order_uid = oe.order_uid AND at.auction_id = oe.auction_id
+    WHERE ss.block_deadline >= {{start_block}}
+        AND ss.block_deadline <= {{end_block}}
+)
+,order_observations AS (
+    SELECT
+        os.auction_id,
+        os.solver,
+        os.tx_hash,
+        os.sell_amount, -- the total amount the user sends
+        os.buy_amount,
+        os.sell_token,
+        os.observed_fee,
+        os.surplus,
+        os.surplus_token,
+        CASE
+            WHEN fp.kind = 'surplus'
+                THEN
+                    CASE
+                        WHEN os.kind = 'sell'
+                            THEN
+                                CASE
+                                    WHEN fp.max_volume_factor = 1 -- this is done to avoid a division by zero bug
+                                        THEN fp.surplus_factor / (1 - fp.surplus_factor) * surplus
```

Comment:
Just something to keep in mind. We should never set surplus_factor = 1. This would break this step, but more importantly, it would make it impossible to be able to compute what the protocol fee is as the order would receive zero surplus in that case.

Reply:
Yes. I did have an additional case for that at some point but it is indeed not possible to compute anything in that case. I will add a comment to make clear what is going on.